tactics/declare.mli

```diff
@@ -157,3 +243,72 @@ module Internal : sig
   val objVariable : unit Libobject.Dyn.tag
 
 end
+
+(* Intermediate step necessary to delegate the future.
+ * Both access the current proof state. The former is supposed to be
+ * chained with a computation that completed the proof *)
+type closed_proof_output
+
+(** Requires a complete proof. *)
+val return_proof : Proof.t -> closed_proof_output
+
+(** An incomplete proof is allowed (no error), and a warn is given if
+   the proof is complete. *)
+val return_partial_proof : Proof.t -> closed_proof_output
+val close_future_proof : feedback_id:Stateid.t -> Proof.t -> closed_proof_output Future.computation -> proof_object
+
+(** [by tac] applies tactic [tac] to the 1st subgoal of the current
+    focused proof.
+    Returns [false] if an unsafe tactic has been used. *)
+val by : unit Proofview.tactic -> Proof.t -> Proof.t * bool
+
+(** Declare abstract constant; will check no evars are possible; *)
+val declare_abstract :
+     name:Names.Id.t
+  -> poly:bool
+  -> kind:Decls.logical_kind
+  -> sign:EConstr.named_context
+  -> secsign:Environ.named_context_val
+  -> opaque:bool
+  -> solve_tac:unit Proofview.tactic
+  -> Evd.evar_map
+  -> EConstr.t
+  -> Evd.side_effects * Evd.evar_map * EConstr.t * EConstr.t list * bool
+
+val build_by_tactic
```

Comment:
What's the plan for this function? It should not belong here.

Reply:
Actually the function itself makes little sense; for example it drops `proof_entry_universe` , it seems to me that all the 3 uses in the codebase are highly suspicious (cc @skyskimmer) 

As of today it has to be here as it deals with entry internals; in my opinion, we should just remove it, and have users go thru the regular declaration path; or provide a very different API.

Reply:
Pseudo-code for this function is:
1. open proof
2. apply a tactic
3. close proof building an entry, don't declare it.
4. disregard any invariants and butcher the entry into some parts of it
so yeah, not very useful in the end; I think that clients would be much better off directly using the proof API itself, not the constant-declaration one.

Reply:
but well, I consider 3 and 4 part of the internal Declare API so this is why the function has to live here for now.

Reply:
For example once obligation code doesn't use its own path, `inline_private_constants` doesn't have to be exported, etc.... and a lot of invariants on sideff expansion can be finally controlled!